File: app/src/main/java/com/appsbygu/kadai_weather_java/activities/WeatherActivity.java
package com.appsbygu.kadai_weather_java.activities;

import android.content.Intent;
import android.databinding.DataBindingUtil;
import android.os.Handler;
import android.os.Message;
import android.support.v7.app.AppCompatActivity;
import android.os.Bundle;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.widget.TextView;

import com.appsbygu.kadai_weather_java.R;
import com.appsbygu.kadai_weather_java.adapters.PinpointLocationAdapter;
import com.appsbygu.kadai_weather_java.databinding.ActivityWeatherBinding;
import com.appsbygu.kadai_weather_java.models.apis.Weathers.PinpointLocation;
import com.appsbygu.kadai_weather_java.models.apis.Weathers.Weather;
import com.appsbygu.kadai_weather_java.services.WeatherAPIService;

import io.realm.RealmList;


public class WeatherActivity extends AppCompatActivity {

    private WeatherAPIService weatherAPIService;
    private Handler handler;
    private ActivityWeatherBinding weatherBinding;
    private RecyclerView.Adapter adapter;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_weather);

        getSupportActionBar().hide();

        Intent intent = getIntent();
        String cityCode = intent.getStringExtra("cityCode");

        weatherBinding = DataBindingUtil.setContentView(this, R.layout.activity_weather);
        weatherAPIService = new WeatherAPIService();
        receiveWeatherInfo(cityCode);

    }

    private void receiveWeatherInfo(String _cityCode) {
        handler = new Handler() {
            public void handleMessage(Message msg){
                Weather weather = (Weather) msg.obj;

                displayTitle(weather);
                displayDiscription(weather);
                displayLocation(weather);
                setPinpointView(weather.getPinpointLocations());
            }
        };
        weatherAPIService.receive_weather_info(handler, _cityCode);
    }

    private void displayTitle(Weather weather) {
        TextView tv = (TextView)findViewById(R.id.titleTV);
        tv.setText(weather.getTitle().trim());
    }

    private void displayDiscription(Weather weather) {
        TextView description = (TextView)findViewById(R.id.descriptionTV);
        description.setText(weather.getDescription().getText());
    }

    private void displayLocation(Weather weather) {
        TextView location = (TextView)findViewById(R.id.locationTV);
        String locationStr = "";
        if(weather.getLocation().getArea() != null) locationStr += weather.getLocation().getArea();
        if(weather.getLocation().getPrefecture() != null) locationStr += " > " + weather.getLocation().getPrefecture();
        if(weather.getLocation().getCity() != null) locationStr += " > " + weather.getLocation().getCity();
        location.setText(locationStr);
    }

    private void setPinpointView(RealmList<PinpointLocation> pinpoints){
        weatherBinding.pinpointRV.setHasFixedSize(true);
        adapter = new PinpointLocationAdapter(pinpoints);
        weatherBinding.pinpointRV.setAdapter(adapter);
        weatherBinding.pinpointRV.setLayoutManager(new LinearLayoutManager(this));
    }
}

File: app/src/main/java/com/appsbygu/kadai_weather_java/activities/WeatherActivity.kt
package com.appsbygu.kadai_weather_java.activities

import android.content.Context
import android.content.Intent
import android.databinding.DataBindingUtil
import android.support.v7.app.AppCompatActivity
import android.os.Bundle
import android.support.v7.widget.LinearLayoutManager
import android.widget.Toast

import com.appsbygu.kadai_weather_java.R
import com.appsbygu.kadai_weather_java.adapters.PinpointLocationAdapter
import com.appsbygu.kadai_weather_java.databinding.ActivityWeatherBinding
import com.appsbygu.kadai_weather_java.models.apis.cities.City
import com.appsbygu.kadai_weather_java.models.apis.weathers.Location
import com.appsbygu.kadai_weather_java.models.apis.weathers.PinpointLocation
import com.appsbygu.kadai_weather_java.models.apis.weathers.Weather
import com.appsbygu.kadai_weather_java.services.WeatherAPIService
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.disposables.CompositeDisposable
import io.reactivex.schedulers.Schedulers

import io.realm.RealmList
import kotlinx.android.synthetic.main.activity_weather.*
import retrofit2.converter.gson.GsonConverterFactory


class WeatherActivity : AppCompatActivity() {

    companion object {

        private const val CITY_CODE = "cityCode"

        fun createIntent(context: Context, city: City): Intent{
            return Intent(context, WeatherActivity::class.java)
                    .putExtra(CITY_CODE, city.id!!.toString())
        }

    }

    private lateinit var weatherBinding: ActivityWeatherBinding
    private val disposables = CompositeDisposable()

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_weather)

        supportActionBar?.hide()
        weatherBinding = DataBindingUtil.setContentView(this, R.layout.activity_weather)

        val cityCode = intent.getStringExtra("cityCode")

        fetchWeatherInfo(cityCode)

    }

    override fun onDestroy() {
        super.onDestroy()
        disposables.dispose()
    }

    private fun fetchWeatherInfo(cityCode: String) {

        val service = WeatherAPIService.getService(GsonConverterFactory.create())
        val dispose = service.weatherInfo(cityCode)
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(this::afterWeatherInfoLoaded, this::weatherInfoLoadFail)
        disposables.add(dispose)
    }

    private fun afterWeatherInfoLoaded(weather: Weather) {
        showTitle(weather)
        showDescription(weather)
        showLocation(weather.location)
        setPinpointView(weather.pinpointLocations)
    }

    private fun weatherInfoLoadFail(error: Throwable) {
        val toast = Toast.makeText(applicationContext, getString(R.string.data_load_error), Toast.LENGTH_LONG)
        toast.show()
    }

    private fun showTitle(weather: Weather) {
        titleTextView.text = weather.title?.trim { it <= ' ' } ?: ""
    }

    private fun showDescription(weather: Weather) {
        descriptionTextView.text = weather.description?.text ?: ""
    }

    private fun showLocation(location: Location?) {
        var locationStr = ""
        location?.area?.let { locationStr += it }
        location?.prefecture?.let { locationStr += " > $it" }
        location?.city?.let { locationStr += " > $it" }
        locationTextView.text = locationStr
    }

    private fun setPinpointView(pinpoints: RealmList<PinpointLocation>?) {
        pinpoints?.let {
            weatherBinding.pinpointRV.setHasFixedSize(true)
            weatherBinding.pinpointRV.adapter = PinpointLocationAdapter(it)
            weatherBinding.pinpointRV.layoutManager = LinearLayoutManager(this)
        }
    }
}
